File: notebooks/generate_sample_data.py
"""
Generates a synthetic support-case dataset for the AI/automation agent DTC
(Days to Close) research project. Designed to mimic the fields described in
the project's Expected Data Sources section, with realistic-ish underlying
relationships baked in:

- Higher complexity/effort -> longer baseline DTC
- Agents reduce DTC, but the effect size depends on agent type vs. complexity:
    * Routing agents help a little across the board (faster triage)
    * Knowledge agents help more on medium complexity / KB-rich areas
    * Scoping/Troubleshooting agents help the most on HIGH complexity cases
    * Blended (multiple agents across lifecycle) generally outperforms any
      single agent, especially on high-priority/high-complexity cases
- Quality metrics (CSAT, reopen rate, SLA attainment) are correlated with DTC
  and automation success so the "maintain/improve quality" angle has signal.

Output: ../data/sample_support_cases.csv
"""

import numpy as np
import pandas as pd
from datetime import timedelta

RNG = np.random.default_rng(42)
N = 1500

product_areas = ["Networking", "Storage", "Identity", "Compute", "Data Platform", "Security", "Collaboration"]
issue_types = ["Bug", "Configuration", "Performance", "How-To", "Outage", "Feature Gap"]
severities = ["Sev-A", "Sev-B", "Sev-C", "Sev-D"]
complexities = ["Low", "Medium", "High"]
effort_levels = ["Low", "Medium", "High"]
agent_types = ["None", "Routing", "Knowledge", "Scoping", "CaseHealth", "Blended"]
lifecycle_stages = ["Creation", "Scoping/Troubleshooting", "Closure"]

agent_probs = [0.30, 0.15, 0.15, 0.15, 0.10, 0.15]  # ~70% of cases have some agent touch

rows = []
start_date = pd.Timestamp("2024-01-01")

for i in range(1, N + 1):
    severity = RNG.choice(severities, p=[0.10, 0.30, 0.40, 0.20])
    complexity = RNG.choice(complexities, p=[0.35, 0.40, 0.25])
    effort = RNG.choice(effort_levels, p=[0.40, 0.35, 0.25])
    product_area = RNG.choice(product_areas)
    issue_type = RNG.choice(issue_types)
    agent_type = RNG.choice(agent_types, p=agent_probs)

    # ---- Baseline DTC driven by severity/complexity/effort ----
    sev_base = {"Sev-A": 3, "Sev-B": 6, "Sev-C": 10, "Sev-D": 14}[severity]
    complexity_add = {"Low": 0, "Medium": 4, "High": 9}[complexity]
    effort_add = {"Low": 0, "Medium": 3, "High": 7}[effort]
    baseline_dtc = sev_base + complexity_add + effort_add + RNG.normal(0, 2)

    # ---- Agent effect (reduction in DTC, in days) ----
    if agent_type == "None":
        reduction = 0
    elif agent_type == "Routing":
        reduction = RNG.normal(1.0, 0.5)
    elif agent_type == "Knowledge":
        reduction = RNG.normal(2.0, 0.7) + (1.5 if complexity == "Medium" else 0)
    elif agent_type == "Scoping":
        reduction = RNG.normal(2.5, 0.8) + (3.5 if complexity == "High" else 0.5)
    elif agent_type == "CaseHealth":
        reduction = RNG.normal(1.5, 0.6)
    else:  # Blended
        reduction = RNG.normal(3.5, 1.0) + (4.0 if complexity == "High" else 1.5) + (1.0 if severity in ("Sev-A", "Sev-B") else 0)

    reduction = max(reduction, 0)
    dtc = max(baseline_dtc - reduction, 0.5)

    # ---- Automation / intervention detail ----
    automation_success = RNG.random() < (0.55 if agent_type == "None" else 0.85)
    num_actions = RNG.integers(1, 4) if agent_type == "None" else RNG.integers(2, 9)
    handoff_hours = round(abs(RNG.normal(6, 4)), 1) if agent_type != "None" else np.nan
    lifecycle_stage_touched = RNG.choice(lifecycle_stages) if agent_type != "None" else None

    # ---- Quality / outcome metrics (correlated with DTC & automation) ----
    first_response_hours = round(max(RNG.normal(4 if agent_type != "None" else 7, 2), 0.2), 1)
    escalation_flag = RNG.random() < (0.25 if complexity == "High" and agent_type == "None" else 0.08)
    reopen_flag = RNG.random() < (0.15 if not automation_success else 0.05)
    resolution_accuracy = round(np.clip(RNG.normal(0.9 if automation_success else 0.75, 0.08), 0, 1), 2)
    csat = round(np.clip(RNG.normal(4.3 if automation_success and dtc < baseline_dtc else 3.6, 0.6), 1, 5), 1)
    sla_target_days = {"Sev-A": 5, "Sev-B": 8, "Sev-C": 12, "Sev-D": 18}[severity]
    sla_met = dtc <= sla_target_days
    repeat_contact_flag = RNG.random() < (0.12 if not automation_success else 0.04)

    # ---- Knowledge / ops context ----
    kb_article_used = RNG.random() < (0.7 if agent_type in ("Knowledge", "Blended") else 0.35)
    kb_coverage_score = round(np.clip(RNG.normal(0.75 if kb_article_used else 0.4, 0.15), 0, 1), 2)
    engineer_workload = RNG.integers(3, 25)
    queue_depth = RNG.integers(5, 80)

    open_date = start_date + timedelta(days=int(RNG.integers(0, 500)), hours=int(RNG.integers(0, 24)))
    close_date = open_date + timedelta(days=float(dtc))

    rows.append({
        "case_id": f"CASE-{100000 + i}",
        "open_date": open_date,
        "close_date": close_date,
        "days_to_close": round(dtc, 2),
        "severity": severity,
        "product_area": product_area,
        "issue_type": issue_type,
        "complexity": complexity,
        "effort_level": effort,
        "escalation_flag": escalation_flag,
        "reopen_flag": reopen_flag,
        "agent_type": agent_type,
        "lifecycle_stage_touched": lifecycle_stage_touched,
        "handoff_time_hours": handoff_hours,
        "num_actions": num_actions,
        "automation_success": automation_success if agent_type != "None" else np.nan,
        "first_response_time_hours": first_response_hours,
        "resolution_accuracy": resolution_accuracy,
        "csat_score": csat,
        "sla_target_days": sla_target_days,
        "sla_met": sla_met,
        "repeat_contact_flag": repeat_contact_flag,
        "kb_article_used": kb_article_used,
        "kb_coverage_score": kb_coverage_score,
        "engineer_workload": engineer_workload,
        "queue_depth": queue_depth,
    })

df = pd.DataFrame(rows)
out_path = "../data/sample_support_cases.csv"
df.to_csv(out_path, index=False)
print(f"Wrote {len(df)} rows to {out_path}")
print(df.groupby("agent_type")["days_to_close"].mean().sort_values())

File: src/generate_sample_data.js
/**
 * Generates a synthetic support-case dataset for the AI/automation agent DTC
 * (Days to Close) research project. See generate_sample_data.py for the full
 * rationale/comments on the underlying relationships baked into the data.
 *
 * Output: ../data/sample_support_cases.csv
 */
const fs = require("fs");
const path = require("path");

// ---- simple seeded RNG (mulberry32) for reproducibility ----
function mulberry32(seed) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
const rand = mulberry32(42);

function randNormal(mean = 0, sd = 1) {
  // Box-Muller transform
  const u1 = Math.max(rand(), 1e-9);
  const u2 = rand();
  const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  return mean + z * sd;
}
function randInt(min, maxExclusive) {
  return Math.floor(rand() * (maxExclusive - min)) + min;
}
function choice(arr, probs) {
  if (!probs) return arr[randInt(0, arr.length)];
  const r = rand();
  let cum = 0;
  for (let i = 0; i < arr.length; i++) {
    cum += probs[i];
    if (r <= cum) return arr[i];
  }
  return arr[arr.length - 1];
}
function clip(v, lo, hi) {
  return Math.min(Math.max(v, lo), hi);
}

const N = 1500;
const productAreas = ["Networking", "Storage", "Identity", "Compute", "Data Platform", "Security", "Collaboration"];
const issueTypes = ["Bug", "Configuration", "Performance", "How-To", "Outage", "Feature Gap"];
const severities = ["Sev-A", "Sev-B", "Sev-C", "Sev-D"];
const complexities = ["Low", "Medium", "High"];
const effortLevels = ["Low", "Medium", "High"];
const agentTypes = ["None", "Routing", "Knowledge", "Scoping", "CaseHealth", "Blended"];
const lifecycleStages = ["Creation", "Scoping/Troubleshooting", "Closure"];
const agentProbs = [0.3, 0.15, 0.15, 0.15, 0.1, 0.15];

const sevBase = { "Sev-A": 3, "Sev-B": 6, "Sev-C": 10, "Sev-D": 14 };
const slaTarget = { "Sev-A": 5, "Sev-B": 8, "Sev-C": 12, "Sev-D": 18 };
const complexityAdd = { Low: 0, Medium: 4, High: 9 };
const effortAdd = { Low: 0, Medium: 3, High: 7 };

const startDate = new Date("2024-01-01T00:00:00Z");
const rows = [];

for (let i = 1; i <= N; i++) {
  const severity = choice(severities, [0.1, 0.3, 0.4, 0.2]);
  const complexity = choice(complexities, [0.35, 0.4, 0.25]);
  const effort = choice(effortLevels, [0.4, 0.35, 0.25]);
  const productArea = choice(productAreas);
  const issueType = choice(issueTypes);
  const agentType = choice(agentTypes, agentProbs);

  const baselineDtc = sevBase[severity] + complexityAdd[complexity] + effortAdd[effort] + randNormal(0, 2);

  let reduction;
  switch (agentType) {
    case "None":
      reduction = 0;
      break;
    case "Routing":
      reduction = randNormal(1.0, 0.5);
      break;
    case "Knowledge":
      reduction = randNormal(2.0, 0.7) + (complexity === "Medium" ? 1.5 : 0);
      break;
    case "Scoping":
      reduction = randNormal(2.5, 0.8) + (complexity === "High" ? 3.5 : 0.5);
      break;
    case "CaseHealth":
      reduction = randNormal(1.5, 0.6);
      break;
    default: // Blended
      reduction =
        randNormal(3.5, 1.0) +
        (complexity === "High" ? 4.0 : 1.5) +
        (severity === "Sev-A" || severity === "Sev-B" ? 1.0 : 0);
  }
  reduction = Math.max(reduction, 0);
  const dtc = Math.max(baselineDtc - reduction, 0.5);

  const automationSuccess = agentType === "None" ? null : rand() < 0.85;
  const automationAttemptSuccess = agentType === "None" ? rand() < 0.55 : automationSuccess;
  const numActions = agentType === "None" ? randInt(1, 4) : randInt(2, 9);
  const handoffHours = agentType !== "None" ? Math.round(Math.abs(randNormal(6, 4)) * 10) / 10 : "";
  const lifecycleStageTouched = agentType !== "None" ? choice(lifecycleStages) : "";

  const firstResponseHours = Math.round(Math.max(randNormal(agentType !== "None" ? 4 : 7, 2), 0.2) * 10) / 10;
  const escalationFlag = rand() < (complexity === "High" && agentType === "None" ? 0.25 : 0.08);
  const reopenFlag = rand() < (!automationAttemptSuccess ? 0.15 : 0.05);
  const resolutionAccuracy = Math.round(clip(randNormal(automationAttemptSuccess ? 0.9 : 0.75, 0.08), 0, 1) * 100) / 100;
  const csat = Math.round(clip(randNormal(automationAttemptSuccess && dtc < baselineDtc ? 4.3 : 3.6, 0.6), 1, 5) * 10) / 10;
  const slaMet = dtc <= slaTarget[severity];
  const repeatContactFlag = rand() < (!automationAttemptSuccess ? 0.12 : 0.04);

  const kbArticleUsed = rand() < (agentType === "Knowledge" || agentType === "Blended" ? 0.7 : 0.35);
  const kbCoverageScore = Math.round(clip(randNormal(kbArticleUsed ? 0.75 : 0.4, 0.15), 0, 1) * 100) / 100;
  const engineerWorkload = randInt(3, 25);
  const queueDepth = randInt(5, 80);

  const openDate = new Date(startDate.getTime() + randInt(0, 500) * 86400000 + randInt(0, 24) * 3600000);
  const closeDate = new Date(openDate.getTime() + dtc * 86400000);

  rows.push({
    case_id: `CASE-${100000 + i}`,
    open_date: openDate.toISOString(),
    close_date: closeDate.toISOString(),
    days_to_close: Math.round(dtc * 100) / 100,
    severity,
    product_area: productArea,
    issue_type: issueType,
    complexity,
    effort_level: effort,
    escalation_flag: escalationFlag,
    reopen_flag: reopenFlag,
    agent_type: agentType,
    lifecycle_stage_touched: lifecycleStageTouched,
    handoff_time_hours: handoffHours,
    num_actions: numActions,
    automation_success: agentType === "None" ? "" : automationSuccess,
    first_response_time_hours: firstResponseHours,
    resolution_accuracy: resolutionAccuracy,
    csat_score: csat,
    sla_target_days: slaTarget[severity],
    sla_met: slaMet,
    repeat_contact_flag: repeatContactFlag,
    kb_article_used: kbArticleUsed,
    kb_coverage_score: kbCoverageScore,
    engineer_workload: engineerWorkload,
    queue_depth: queueDepth,
  });
}

const headers = Object.keys(rows[0]);
const csvLines = [headers.join(",")];
for (const row of rows) {
  csvLines.push(headers.map((h) => row[h]).join(","));
}

const outPath = path.join(__dirname, "..", "data", "sample_support_cases.csv");
fs.writeFileSync(outPath, csvLines.join("\n"));
console.log(`Wrote ${rows.length} rows to ${outPath}`);

// quick summary: avg DTC by agent type
const byAgent = {};
for (const r of rows) {
  if (!byAgent[r.agent_type]) byAgent[r.agent_type] = [];
  byAgent[r.agent_type].push(r.days_to_close);
}
for (const [agent, vals] of Object.entries(byAgent)) {
  const avg = vals.reduce((a, b) => a + b, 0) / vals.length;
  console.log(`${agent}: avg DTC = ${avg.toFixed(2)} (n=${vals.length})`);
}
